art/attacks/poisoning/bullseye_polytope_attack.py

```diff
@@ -0,0 +1,422 @@
+        elif opt_method == 'signedadam':
+            logger.info("Using Signed Adam to craft poison samples")
+            optimizer = SignedAdam(poison_batch.parameters(), lr=self.learning_rate, betas=(self.momentum, 0.999))
+        elif opt_method == 'adam':
+            logger.info("Using Adam to craft poison samples")
+            optimizer = torch.optim.Adam(poison_batch.parameters(), lr=self.learning_rate, betas=(self.momentum, 0.999))
+
+        base_tensor_batch = torch.stack(base_tensor_list, 0)
+        base_range01_batch = base_tensor_batch * std + mean
+
+        # Because we have turned on DP for the substitute networks,
+        # the target image's feature becomes random.
+        # We can try enforcing the convex polytope in one of the multiple realizations of the feature,
+        # but empirically one realization is enough.
+        target_feat_list = []
+        # Coefficients for the convex combination.
+        # Initializing from the coefficients of last step gives faster convergence.
+        s_init_coeff_list = []
+        n_poisons = len(x)
+        for n, net in enumerate(self.subsistute_networks):
+            # net = net.model
+            # net.eval()
+            # End to end training
+            if self.endtoend:
+                block_feats = [feat.detach()
+                               for feat in net.get_activations(x, layer=self.feature_layer, framework=True)]
+                target_feat_list.append(block_feats)
+                s_coeff = [torch.ones(n_poisons, 1).to(self.estimator.device) / n_poisons for _ in
+                           range(len(block_feats))]
+            else:
+                target_feat_list.append(net.get_activations(x, layer=self.feature_layer, framework=True).detach())
+                s_coeff = torch.ones(n_poisons, 1).to(self.estimator.device) / n_poisons
+
+            s_init_coeff_list.append(s_coeff)
+
+        for ite in trange(self.max_iter):
+            if ite % self.decay_iter == 0 and ite != 0:
+                for param_group in optimizer.param_groups:
+                    param_group['lr'] *= self.decay_coeff
+                print("%s Iteration %d, Adjusted lr to %.2e" % (time.strftime("%Y-%m-%d %H:%M:%S"), ite,
+                                                                self.learning_rate))
+
+            poison_batch.zero_grad()
+            total_loss = loss_from_center(self.subsistute_networks, target_feat_list, poison_batch, self.net_repeat,
+                                          self.endtoend, self.feature_layer)
+            total_loss.backward()
+
+            optimizer.step()
+
+            # clip the perturbations into the range
+            perturb_range01 = torch.clamp((poison_batch.poison.data - base_tensor_batch) * std,
+                                          -self.epsilon,
+                                          self.epsilon)
+            perturbed_range01 = torch.clamp(base_range01_batch.data + perturb_range01.data, 0, 1)
```

Comment:
Can we use here `estimator.clip_values` for clipping?

Reply:
Done.

Reply:
Done.